Overhaul power-up sequence document (#317)
gbdev/pandocs: src/Power_Up_Sequence.md

# Power-Up Sequence

When the Game Boy is powered up, the CPU starts executing instuctions at address $0000—**not** $0100!
A program called the *boot ROM*, burned inside the CPU, is mapped "over" the cartridge ROM at first, which is responsible for the boot-up animation played before control is handed over to the cartridge's ROM.

There exist 8 different known official boot ROMs:

Name | Size (bytes) | Notes
-----|--------------|------------------------------------------------------------
DMG0 | 256          | Blinks on failed checks, no ®
DMG  | 256          | |
MGB  | 256          | One-byte difference to DMG
SGB  | 256          | Only forwards logo to SGB BIOS, performs no checks
SGB2 | 256          | Same difference to SGB than between MGB and DMG
CGB0 | 256 + 1792   | Does not init [wave RAM](<#FF30-FF3F - Wave Pattern RAM>)
CGB  | 256 + 1792   | Split in two parts, with the cartridge header in the middle
AGB  | 256 + 1792   | Fixes "logo TOCTTOU"

> A disassembly of all of them [is available online](https://github.com/ISSOtm/gb-bootroms).

Author: I've been advised to avoid short, one-word links for being smaller targets and not descriptive/accessible.

> A disassembly of all of them [is available online](https://github.com/ISSOtm/gb-bootroms).

Reviewer: You can link the whole phrase if you prefer

Additionally, fauly or dirty connections can cause the data read to be corrupted, resulting in a jumbled-up logo.

*Once the logo has finished scrolling*, the boot ROM plays the famous "ba-ding!" sound, and reads the logo **again**, this time comparing it to a copy it stores.
Then, it also computes the header checksum, and compares it to [the checksum stored in the header](<#014D - Header Checksum>).
If either of these checks fail, the boot ROM **locks up**, and control is never passed to the cartridge ROM.

Finally, the boot ROM writes to the `BANK` register at $FF50, which unmaps the boot ROM.
The `ldh [$FF50], a` instruction being located at $00FE (and being two bytes long), [the first instruction executed from the cartridge ROM is at $0100](<#0100-0103 - Entry Point>).

Since the A register is used to write to $FF50, its value is passed to the cartridge ROM; the only difference between the DMG and MGB boot ROMs is that the former writes $01, and the latter uses $FF.

### DMG0

The DMG0 is a rare "early bird" variant of the DMG boot ROM present in few early DMGs.
The behavior of the boot ROM is globally the same, but significant portions of the code have been rearranged.

> Interestingly, the DMG0 boot ROM performs both the logo and checksum checks before displaying anything, and if either fails, the screen is made to blink while the boot ROM locks up, alternating between solid white and solid black.

Reviewer: *if either fails, the screen is made to blink while the boot ROM locks up* looks like a dependent clause so no comma before the connecting and.

If either verification fails, the screen is made to blink while the boot ROM locks up, alternating between solid white and solid black.

The DMG0 boot ROM also lacks the ® symbol next to the Nintendo logo.

> Like the DMG, the DMG0 boot ROM writes $01 to $FF50.

Author: Yes, do you want it to be said again there?

> Like the DMG, the DMG0 boot ROM writes $01 to $FF50.

Reviewer: I'm not sure why it is needed to say here. If you are listing differences, then this is not necessary (as it is the same)

## Super Game Boy (SGB, SGB2)

These boot ROMs are fairly unique in that they do *not* perform header checks.
Instead, they set up the Nintendo logo in VRAM from the header just like the monochrome boot ROMs, but then they send the entire header to the SGB BIOS via [the standard packet-transferring procedure](<#Command Packet Transfers>), using packet header bytes $F1, $F3, $F5, $F7, $F9, and $FB, in that order.
(These packet IDs are otherwise invalid and never used in regular SGB operation, though it seems that not all SGB BIOS revisions filter them out.)

The boot ROM then unmaps itself and hands off execution to the cartridge ROM without performing any checks.

> The SGB BIOS, the program running on the SNES, actually verifies the Nintendo logo and header checksum itself, and if either fails, it itself locks up, and repeatedly resets the SGB CPU within the cartridge.

Reviewer: Same as previous. The phrasing also sounds a bit convoluted

The way the packet-sending routine works makes transferring a set bit *one cycle* faster than transferring a reset bit.
This means that the time taken by the SGB boot ROMs *depends on the cartridge's header*!
The complexity of this is compounded by the fact that the boot ROM waits for 4 VBlanks after transferring each packet, mostly but not entirely grouping the timings.

## Color models (CGB0, CGB, AGB)

The color boot ROMs are much more complicated, notably because of the compatibility behavior.

First, the boot ROM is larger, as indicated above: 2048 bytes total.
It still has to be mapped starting at $0000, since this is where the CPU starts, but it must also access the cartridge header at $0100-014F.
Thus, the boot ROM is actually split in two parts, a $0000-00FF one, and a $0200-08FF one.

First, the boot ROMs unpack the Nintendo logo to VRAM like the monochrome models, likely for compatibility, and copies the logo to a buffer in HRAM at the same time.
(It is speculated that HRAM was used due to it being embedded within the CPU, unlike WRAM, so that it couldn't be tampered with.)

> Then, the logo is read *again*, but this time, it is decompressed without upscaling, yielding the much smaller logo placed below the big "GAME BOY" one.

Author: It's upscaled "like the monochrome models".
```suggestion
Then, the logo is read *again*, but this time, unlike monochrome models, it is decompressed without upscaling, yielding the much smaller logo placed below the big "GAME BOY" one.
```

The boot ROM then sets up compatibility palettes, as described further below, and plays the logo animation with the "ba-ding!" sound.

During the logo animation, and if bit 7 of [the CGB compatibility byte](<#0143 - CGB Flag>) is reset (indicating a monochrome-only game), the user is allowed to pick a palette to override the one chosen for compatibility.
Each new choice prevents the animation from ending for 30 frames, potentially delaying the checks and fade-out.

Then, like the monochrome boot ROMs, the header logo is checked *from the buffer in HRAM*, and the header checksum is verified.

> For some reason, however, only the first half of the logo is checked, despite the full logo being present in the HRAM buffer.

Author: I wanted to express both that this largely defeats the purpose of the hardened check ("however"), and that the rationale is hard to guess ("for some reason"). Is it too much, though?

> For some reason, however, only the first half of the logo is checked, despite the full logo being present in the HRAM buffer.

Reviewer: For some reason, however is definitely wrong. If you want to express the unknown reason you could go with "For unknown reasons"

> Finally, the boot ROM fades all BG palettes to white, and sets up compatibility.

Author: Well, it does set up the hardware to go into compatibility mode (`OPRI`, `KEY0` and all that).

     (So, if the letter was found in the first row, the index is unchanged; if it's found in the second row, it's increased by 14, and so on.)

The resulting palettes ID is used to pick 3 palettes out of a table via a fairly complex mechanism.
The user can override this choice using certain button combinations during the logo animation; some of these manual choices are identical to auto-colorizations, [but others are unique](https://tcrf.net/Notes:Game_Boy_Color_Bootstrap_ROM#Manual_Select_Palette_Configurations).

::: tip

A table of checksums (and tie-breaker fourth letters when applicable) and the corresponding palettes can be found [on TCRF](https://tcrf.net/Notes:Game_Boy_Color_Bootstrap_ROM#Assigned_Palette_Configurations).

:::

If the ID is either $43 or $58, then the Nintendo logo's tilemap is written to VRAM.
This is intended for games that perform some kind of animation with the Nintendo logo; it suddenly appears in the middle of the screen, though, so it may look better for homebrew not to use this mechanism.

### Scrapped palette switching menu

Remnants of a functionality designed to allow switching the CGB palettes while the game is running exist in the CGB CPU.

> TODO

Reviewer: Please avoid leaving TODOs around. Remove/comment/open an issue for WIP chapters

This requires custom hardware in the cartridge, however, and is made difficult by the timing and order of the reads varying greatly between boot ROMs.
Some implementations use a custom mapper, others use a capacitor holding some of the address lines to redirect reads to a separate region of ROM containing the modified logo.

The other way is Game Boy Color (and Advance) exclusive: for some reason, the boot ROM copies the full logo into HRAM, but only compares the first half.
Thus, a logo whose top half is correct but not the bottom half will get a pass from the CGB boot ROM.
Strangely, despite correcting the TOCTTOU vulnerability, the CGB-AGB boot ROM does *not* fix this mistake.

## Console state after boot ROM hand-off

Regardless of the console you intend for your game to run on, it is prudent to rely on as little of the following as possible, barring what is mentioned elsewhere in this documentation to detect which system you are running on.
This ensures maximum compatibility, both across consoles and cartridges (especially flashcarts, which typically run their own menu code before your game), increases reliability, and is generally considered good practice.

> Due to the high volatility of the information presented below (due to the complexity of some of the boot ROM behaviors), what is documented below but not elsewhere might contain errors.

Reviewer: Needs a rephrasing

Rely on it at your own risk.

### Common remarks

The console's WRAM and HRAM are random on power-up.
[Different models tend to exhibit different patterns](https://twitter.com/CasualPkPlayer/status/1409752977812852736?s=19), but they are random nonetheless, even depending on factors such as the ambient teperature.
Besides, turning the system off and on again has proven reliable enough [to carry over RAM from one game to another](https://www.youtube.com/watch?v=xayxmTLljr8), so it's not a good idea to rely on it at all.

> Emulation of uninitialized RAM is spotty: some emulators fill RAM with a constant on startup (typically $00 or $FF), some emulators fully randomize RAM, and others attempt to reproduce the patterns observed on hardware.

Reviewer: ```suggestion
Emulation of uninitialized RAM is inconsistent: some emulators fill RAM with a constant on startup (typically $00 or $FF), some emulators fully randomize RAM, and others attempt to reproduce the patterns observed on hardware.
```

# Power-Up Sequence

> When the Game Boy is powered up, the CPU starts executing instuctions at address $0000—**not** $0100!

Reviewer: Why is it needed to enforce this much (bold, exclamation mark) that it doesn't start at $0100? Isn't "contrary to popul belief" or somehting like that enough? Why even mention this?

Author: Because everything else in the Pan Docs, and most other documentation, claims that execution starts at $0100, which *is* valid in "userland" context, but not in boot ROM context. Do you think I should word it that way instead?

Reviewer: Makes sense, yes! (also avoid enforcing it that much)

## Super Game Boy (SGB, SGB2)

These boot ROMs are fairly unique in that they do *not* perform header checks.
Instead, they set up the Nintendo logo in VRAM from the header just like the monochrome boot ROMs, but then they send the entire header to the SGB BIOS via [the standard packet-transferring procedure](<#Command Packet Transfers>), using packet header bytes $F1, $F3, $F5, $F7, $F9, and $FB, in that order.
(These packet IDs are otherwise invalid and never used in regular SGB operation, though it seems that not all SGB BIOS revisions filter them out.)

The boot ROM then unmaps itself and hands off execution to the cartridge ROM without performing any checks.
The SGB BIOS, the program running on the SNES, actually verifies the Nintendo logo and header checksum itself.
If either verification fails, the BIOS itself locks up, repeatedly resetting the SGB CPU within the cartridge.

As the DMG and MGB boot ROMs, the SGB and SGB2 boot ROMs write $01 and $FF respectively to $FF50, and this is also the only difference between these two boot ROMs.

The way the packet-sending routine works makes transferring a set bit *one cycle* faster than transferring a reset bit.

> This means that the time taken by the SGB boot ROMs *depends on the cartridge's header*!

Reviewer: ```suggestion
This means that the time taken by the SGB boot ROMs *depends on the cartridge's header*.
```

> The complexity of this is compounded by the fact that the boot ROM waits for 4 VBlanks after transferring each packet, mostly but not entirely grouping the timings.

Reviewer: Complexity of what? Of understanding the timing?

Author: The timing itself, given that the boot ROM then polls LY, so there is some sort of grouping of timings as well, I believe.

## Color models (CGB0, CGB, AGB)

The color boot ROMs are much more complicated, notably because of the compatibility behavior.

First, the boot ROM is larger, as indicated above: 2048 bytes total.
It still has to be mapped starting at $0000, since this is where the CPU starts, but it must also access the cartridge header at $0100-014F.
Thus, the boot ROM is actually split in two parts, a $0000-00FF one, and a $0200-08FF one.

> First, the boot ROMs unpack the Nintendo logo to VRAM like the monochrome models, likely for compatibility, and copies the logo to a buffer in HRAM at the same time.

Reviewer: First again (?)

Author: The first "First" is the first of the two items, size & memory organization; the second "First" is the first *sub*-item of the second item, behavioral differences. I now realize that I failed to mention the second item at all, so I will restructure this part a bit.

(It is speculated that HRAM was used due to it being embedded within the CPU, unlike WRAM, so that it couldn't be tampered with.)

> Then, the logo is read *again*, but this time, it is decompressed with no resizing, yielding the much smaller logo placed below the big "GAME BOY" one.

Reviewer: ```suggestion
Then, the logo is read *again*, this time decompressed with no resizing, yielding the much smaller logo placed below the big "GAME BOY" one.
```

Additionally, the DMG logo tilemap is written if the compatibility requests it.

Like all other boot ROMs, the last thing the color boot ROMs do is hand off execution at the same time as they unmap themselves, though they write $11 instead of $01 or $FF.

### CGB0

Like the DMG0 boot ROM, some early CGBs contain a different boot ROM.
Unlike DMG0 and DMG, the differences between the CGB0 and CGB boot ROM, are very minor, with no change in the layout of the ROM.

The most notable change is that the CGB0 boot ROM does *not* init [wave RAM](<#FF30-FF3F - Wave Pattern RAM>).
This is known to cause, for example, a different title screen music in the game *R-Type*.

The CGB0 boot ROM also writes copies of other variables to some locations in WRAM that are not otherwise run from anywhere.
It is speculated that this may be debug remnants.

### Compatibility palettes

> It is fairly well-known that the Game Boy Color automatically colorizes monochrome-only games.

Reviewer: ```suggestion
The boot ROM is responsible for the colorization feature of monochrome-only games on Game Boy Color.
```

> The boot ROM is mostly responsible for this.

Reviewer: ```suggestion
```

The resulting palettes ID is used to pick 3 palettes out of a table via a fairly complex mechanism.
The user can override this choice using certain button combinations during the logo animation; some of these manual choices are identical to auto-colorizations, [but others are unique](https://tcrf.net/Notes:Game_Boy_Color_Bootstrap_ROM#Manual_Select_Palette_Configurations).

::: tip

A table of checksums (and tie-breaker fourth letters when applicable) and the corresponding palettes can be found [on TCRF](https://tcrf.net/Notes:Game_Boy_Color_Bootstrap_ROM#Assigned_Palette_Configurations).

:::

If the ID is either $43 or $58, then the Nintendo logo's tilemap is written to VRAM.
This is intended for games that perform some kind of animation with the Nintendo logo; it suddenly appears in the middle of the screen, though, so it may look better for homebrew not to use this mechanism.

### Scrapped palette switching menu

Remnants of a functionality designed to allow switching the CGB palettes while the game is running exist in the CGB CPU.

## Stadium 2

Pokémon Stadium 2's "GB Tower" emulator contains a very peculiar boot ROM.

> It can be found at offset $015995F0 in the US release, and is $3F0 bytes long.

Reviewer: Why the length here is in hex?

Author: No particular reason, [I just copied it verbatim](https://github.com/ISSOtm/gb-bootroms/blob/6232573bc6592df17cdbce878c418e79d8355b68/src/stadium2.asm#L796). Is decimal preferable?

Its purpose is unknown.

This boot ROM does roughly the same setup as a regular CGB boot ROM, but writes to $FF50 very early, and said write is followed by a lock-up loop.
Further, the boot ROM contains a valid header, which is mostly blank save for the logo, compatibility flag (which indicates dual compatibility), and header checksum.

## Logo check

> While it may make sense for the boot ROM to at least partially verify the ROM's integrity via the header check, why does it check the logo's integrity separately?

Reviewer: I don't see why this needs to be worded as a direct question to the reader, also supposing known information (it may make sense to at least partially verify? why?)
